### PR TITLE
Basic online compatibility and a lot of other stuff

### PR DESCRIPTION
**Pull request by maybegreat48**
* Added an anticheat bypass
* Fixed noclip
* Added settings serialization
* Refactored the GUI
* Added some protections (does not currently block every crash)
* Notification Service
* Players list

TODO:
* Readd column components

### CODE REVIEW

**Comment by DayibBaba** [comment]
Were the hotkeys broken on the main branch?

File: src/game/features/self/InfiniteAmmo.cpp [hunk 1]
```diff
@@ -11,17 +11,12 @@ namespace YimMenu::Features
...
 		{
-			Hash current_weapon{};
-			int max_ammo{};
-
-			WEAPON::GET_CURRENT_PED_WEAPON(Self::PlayerPed, &current_weapon, false, false, false);
-			WEAPON::GET_MAX_AMMO(Self::PlayerPed, &max_ammo, current_weapon);
-
-			auto current_weapon_ammo_type = WEAPON::_GET_CURRENT_PED_WEAPON_AMMO_TYPE(Self::PlayerPed, WEAPON::_GET_PED_WEAPON_OBJECT(Self::PlayerPed, false));
-			auto current_ammo = WEAPON::GET_PED_AMMO_BY_TYPE(Self::PlayerPed, current_weapon_ammo_type);
+			WEAPON::SET_PED_INFINITE_AMMO(Self::PlayerPed, true, 0);
```
Comment (DayibBaba) [inline src/game/features/self/InfiniteAmmo.cpp:14]: Not sure whether using natives here would be desirable, since they tend to do some extra stuff I'd rather avoid.

File: src/game/frontend/manager/Widgets/LoopedCommand.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ namespace YimMenu
 
 		if (ImGui::IsItemHovered())
 		{
-			ImGui::SetTooltip(std::format("{}\n\nPress shift to assign a hotkey", Command->GetDescription()).data());
+			ImGui::SetTooltip(std::format("{}\n\nPress shift to assign a hotkey", Command->GetDescription()).data()); // TODO: do we need this?
```
Comment (DayibBaba) [inline src/game/frontend/manager/Widgets/LoopedCommand.cpp:50]: Which part are you referring to? The actual description or the button suggestion? Or the fact there is a popup at all?

File: src/game/hooks/GUI/Window.cpp [hunk 3]
```diff
@@ -10,9 +10,6 @@ namespace YimMenu::Hooks
 		if (g_Running)
 			Renderer::WndProc(hwnd, umsg, wparam, lparam);
 
-		if (umsg == WM_KEYUP && wparam == VK_DELETE) // fallback to unload
```
Comment (DayibBaba) [inline src/game/hooks/GUI/Window.cpp:13]: This button was useful when a certain UI change during development resulted in the UI button being unusable.

File: src/game/features/self/InfiniteAmmo.cpp (hunk 1 above)
Comment (DayibBaba) [reply src/game/features/self/InfiniteAmmo.cpp:14]: Figures.

File: src/game/hooks/GUI/Window.cpp (hunk 3 above)
Comment (DayibBaba) [reply src/game/hooks/GUI/Window.cpp:13]: I'm guessing unloading online raises detections concerns.

**Comment by DayibBaba** [comment]
Is the migrationdatanode bs the same here as in gta?

**Comment by DayibBaba** [comment]
> > Is the migrationdatanode bs the same here as in gta?
> 
> Yes. The P2Cs apparently ported that from YimMenu

Figures

**Comment by Hydra-Source** [comment]
Please split the PR next time, title is "Basic online compatibility" but i see you refactored the GUI and did other stuff that has nothing to do with online. It would be better if you do them one pr at a time so we can also review them easily and rapidly. That way we can avoid unnecessary code bugs and issues.

File: src/game/rdr/Player.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+		return m_Handle->GetGamerInfo();
+	}
+
+	Entity Player::GetPed()
+	{
+		if (!m_Handle || !m_Handle->m_PlayerInfo)
+			return nullptr;
+
+		return m_Handle->m_PlayerInfo->m_Ped;
```
Comment (DayibBaba) [inline src/game/rdr/Player.cpp:47]: Does this really return an Entity?

File: src/game/features/Features.cpp [hunk 5]
```diff
@@ -17,19 +17,21 @@ namespace YimMenu
 
 		if (PED::IS_PED_IN_ANY_VEHICLE(Self::PlayerPed, true))
 			Self::Veh = PED::GET_VEHICLE_PED_IS_IN(Self::PlayerPed, true);
+		else
+			Self::Veh = 0;
 
 		if (PED::IS_PED_ON_MOUNT(Self::PlayerPed))
 			Self::Mount = PED::GET_MOUNT(Self::PlayerPed);
-		
-		if (!ENTITY::DOES_ENTITY_EXIST(Self::Mount) && ENTITY::DOES_ENTITY_EXIST(PED::_GET_LAST_MOUNT(Self::PlayerPed)))
```
Comment (DayibBaba) [inline src/game/features/Features.cpp:24]: This way we lose the handle to the last mount if there was any.

File: src/core/frontend/manager/Submenu.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+				color.w -= 0.5;
+
+				if (category == GetActiveCategory())
+					ImGui::PushStyleColor(ImGuiCol_Button, color);
+
+				if (ImGui::Button(category->m_Name.data(), ImVec2(75, 35)))
+				{
+					SetActiveCategory(category);
+				}
+				if (category == GetActiveCategory())
```
Comment (Hydra-Source) [inline src/core/frontend/manager/Submenu.cpp:35]: Will not pop color if a new category is selected

File: src/core/frontend/manager/UIManager.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		}
+		ImGui::EndChild();
+		
+		ImGui::Text("HorseMenu");
+
+		ImGui::SetCursorPos(ImVec2(pos.x + 130, pos.y));
+
+		if (ImGui::BeginChild("##minisubmenus", ImVec2(0, 50), true, ImGuiWindowFlags_NoScrollbar))
+		{
+			GetActiveSubmenu()->DrawCategorySelectors();
```
Comment (Hydra-Source) [inline src/core/frontend/manager/UIManager.cpp:40]: Check for nullptr

File: src/core/frontend/manager/UIManager.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		}
+		ImGui::EndChild();
+
+		ImGui::SetCursorPos(ImVec2(pos.x + 130, pos.y + 60));
+
+		if (ImGui::BeginChild("##options", ImVec2(0, 0), true))
+		{
+			if (m_OptionsFont)
+				ImGui::PushFont(m_OptionsFont);
+			GetActiveSubmenu()->Draw();
```
Comment (Hydra-Source) [inline src/core/frontend/manager/UIManager.cpp:50]: Check for nullptr

File: src/core/frontend/manager/UIManager.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		{
+			GetActiveSubmenu()->DrawCategorySelectors();
+		}
+		ImGui::EndChild();
+
+		ImGui::SetCursorPos(ImVec2(pos.x + 130, pos.y + 60));
+
+		if (ImGui::BeginChild("##options", ImVec2(0, 0), true))
+		{
+			if (m_OptionsFont)
```
Comment (Hydra-Source) [inline src/core/frontend/manager/UIManager.cpp:48]: Is this really needed?

File: src/game/features/self/Suicide.cpp [hunk 10]
```diff
@@ -0,0 +1,28 @@
+#include "core/commands/Command.hpp"
+#include "game/features/Features.hpp"
+#include "game/rdr/Natives.hpp"
+
+#include "game/rdr/Nodes.hpp"
```
Comment (Hydra-Source) [inline src/game/features/self/Suicide.cpp:5]: Why are these included?

File: src/game/frontend/Menu.hpp [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace YimMenu::Menu
+{
+	extern void Init();
+	extern void SetupStyle();
+	extern void SetupFonts();
+
+	namespace Font
+	{
+		inline ImFont* sm_DefaultFont   = nullptr;
```
Comment (Hydra-Source) [inline src/game/frontend/Menu.hpp:12]: sm? static member? Member?

File: src/game/frontend/items/CommandItem.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+			ImGui::BulletText("Hover over the command name to change its hotkey");
+			ImGui::BulletText("Press any registered key to remove");
+			ImGui::Separator();
+
+			HotkeySetter(m_Command->GetHash()).Draw();
+
+			
+			ImGui::Spacing();
+			if (ImGui::Button("Close") || ((!ImGui::IsWindowHovered() && !ImGui::IsAnyItemHovered()) && ImGui::IsMouseClicked(ImGuiMouseButton_Left)))
+				ImGui::CloseCurrentPopup();
```
Comment (Hydra-Source) [inline src/game/frontend/items/CommandItem.cpp:51]: Doesn't imgui alrady close the popup when clicked outside?

File: src/game/hooks/Info/AssignPhysicalIndex.cpp [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+	{
+		if (newIndex == -1)
+		{
+			Players::HandlePlayerLeave(player);
+			BaseHook::Get<Info::AssignPhysicalIndex, DetourHook<decltype(&Info::AssignPhysicalIndex)>>()->Original()(mgr, player, newIndex);
+		}
+		else
+		{
+			BaseHook::Get<Info::AssignPhysicalIndex, DetourHook<decltype(&Info::AssignPhysicalIndex)>>()->Original()(mgr, player, newIndex);
+			Players::HandlePlayerJoin(player);
```
Comment (Hydra-Source) [inline src/game/hooks/Info/AssignPhysicalIndex.cpp:18]: Not a fool-proof way of storing players. Use parsing from CNetworkPlayerMgr

File: src/main.cpp [hunk 14]
```diff
@@ -40,17 +44,16 @@ namespace YimMenu
...
 		ScriptMgr::AddScript(std::make_unique<Script>(&BlockControlsForUI));
 
-		g_HotkeySystem.RegisterCommands();
-
 		while (g_Running)
 		{
-			//Needed incase UI is malfunctioning or for emergencies
-			if (GetAsyncKeyState(VK_DELETE) & 0x8000)
+			// Needed incase UI is malfunctioning or for emergencies
+			if (GetAsyncKeyState(VK_DELETE) & 0x8000 && !NETWORK::NETWORK_IS_SESSION_STARTED())
```
Comment (Hydra-Source) [inline src/main.cpp:50]: Atleast use the ptr if you insist unloading in online should absolutely be forbidden

File: src/game/backend/Players.cpp [hunk 15]
```diff
@@ -0,0 +1,26 @@
+#include "Players.hpp"
+#include <network/CNetGamePlayer.hpp>
+
+namespace YimMenu
+{
+	void Players::HandlePlayerJoinImpl(CNetGamePlayer* player)
+	{
+		m_Players.insert({ player->GetName(), player });
```
Comment (Hydra-Source) [inline src/game/backend/Players.cpp:8]: Use [] = instead to override

File: src/game/hooks/Info/AssignPhysicalIndex.cpp [hunk 16]
```diff
@@ -0,0 +1,21 @@
+#include "core/hooking/DetourHook.hpp"
+#include "game/hooks/Hooks.hpp"
+#include "game/pointers/Pointers.hpp"
+#include "game/backend/Players.hpp"
+
+namespace YimMenu::Hooks
+{
+	void Info::AssignPhysicalIndex(void* mgr, CNetGamePlayer* player, uint8_t newIndex)
+	{
+		if (newIndex == -1)
```
Comment (Hydra-Source) [inline src/game/hooks/Info/AssignPhysicalIndex.cpp:10]: unsigned == -1 ?

File: src/game/hooks/Info/AssignPhysicalIndex.cpp [hunk 17]
```diff
@@ -0,0 +1,21 @@
+#include "core/hooking/DetourHook.hpp"
+#include "game/hooks/Hooks.hpp"
+#include "game/pointers/Pointers.hpp"
+#include "game/backend/Players.hpp"
+
+namespace YimMenu::Hooks
+{
+	void Info::AssignPhysicalIndex(void* mgr, CNetGamePlayer* player, uint8_t newIndex)
```
Comment (Hydra-Source) [inline src/game/hooks/Info/AssignPhysicalIndex.cpp:8]: New players will not be overridden and old's will not be deleted

File: src/game/frontend/items/CommandItem.cpp (hunk 12 above)
Comment (DayibBaba) [reply src/game/frontend/items/CommandItem.cpp:51]: Not for modals.

**Comment by Hydra-Source** [comment]
Also is hotkeys fixed?

**Comment by Hydra-Source** [comment]
"TODO:

    Readd column components
"?

```c++
class next_column : public option {
    public:
		next_column() {};

		void draw() override {
			ImGui::NextColumn();
		}
	};

	class set_column_offset : public option {
	public:
		set_column_offset(int column, int offset) : m_column(column), m_offset(offset) {};

		void draw() override {
			ImGui::SetColumnOffset(m_column, m_offset);
		}
	private:
		int m_column;
		int m_offset;
	};

	class column : public option {
	public:
		column(int columns) : m_columns(columns) {};

		void add_option(std::shared_ptr<option> box) {
			m_options.push_back(std::move(box));
		}

		void add_next_column() {
			m_options.push_back(std::make_shared<next_column>());
		}

		void add_offset(int column, int offset) {
			m_options.push_back(std::make_shared<set_column_offset>(column, offset));
		
		}

		void draw() override {

			ImGui::Columns(m_columns);

			for (const auto& box : m_options) {
				if (box) {
					box->draw();
				}			
			}
		}

	private:
		int m_columns;
		std::vector<std::shared_ptr<option>> m_options;
	};
	``` This?

File: src/game/rdr/Player.cpp (hunk 4 above)
Comment (Hydra-Source) [reply src/game/rdr/Player.cpp:47]: > It doesn't, but we don't have a ped helper class yet

PTR to handle? You already have it.

File: src/core/frontend/manager/UIManager.cpp (hunk 9 above)
Comment (Hydra-Source) [reply src/core/frontend/manager/UIManager.cpp:48]: > I'd like to isolate `core` and reuse it for other projects like YimMenuV2

I mean the m_OptionsFont nullptr check

File: src/game/rdr/Player.cpp (hunk 4 above)
Comment (Hydra-Source) [reply src/game/rdr/Player.cpp:47]: Wait does this even compile with no warnings/errors?

**Comment by DayibBaba** [comment]
Migrated to #36
